File: real_time_crawler/Python/Callback.py
# This is a simple Sanic web server with a route listening for callbacks on localhost:8080.
# It will print job results to stdout.
import requests
from pprint import pprint
from sanic import Sanic, response


AUTH_TUPLE = ('user', 'pass1')

app = Sanic()


# Define /job_listener endpoint that accepts POST requests.
@app.route('/job_listener', methods=['POST'])
async def job_listener(request):
    try:
        res = request.json
        links = res.get('_links', [])
        for link in links:
            if link['rel'] == 'results':
                # Sanic is async, but requests are synchronous, to fully take
                # advantage of Sanic, use aiohttp.
                res_response = requests.request(
                    method='GET',
                    url=link['href'],
                    auth=AUTH_TUPLE,
                )
                pprint(res_response.json())
                break
    except Exception as e:
        print("Listener exception: {}".format(e))
    return response.json(status=200, body={'status': 'ok'})


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8080)

File: scraper-apis/Nodejs/Callback.js
// Simple express server with an endpoint listening for POST requests
const axios = require('axios');
const express = require('express');
const bodyParser = require('body-parser');

const credentials = {
  auth: {
    username: 'user',
    password: 'pass1'
  },
  headers: {
    'Content-Type': 'application/json',
    'Accept-Encoding': 'gzip,deflate,compress',
  }
}

const app = express();
const port = 8080;

app.use(bodyParser.json());

app.post('/job_listener', (req, res) => {
  const links = req.body._links
  const resultLinks = links.filter(link => link.rel === 'results');

  // create an array of promises from filtered links
  const requests = resultLinks.map(({ href }) => axios.get(href, {...credentials}));

  Promise.all(requests).then(values => {
    console.log(values);
    res.status(200).json({ status: 'ok '});
  })
  .catch(errors => console.log(errors));
});

app.listen(port);
